runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/status/BeamWorkerStatusGrpcService.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.beam.model.pipeline.v1.Endpoints.ApiServiceDescriptor;
+import org.apache.beam.runners.fnexecution.FnService;
+import org.apache.beam.runners.fnexecution.HeaderAccessor;
+import org.apache.beam.runners.fnexecution.status.WorkerStatusClient;
+import org.apache.beam.vendor.grpc.v1p21p0.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A Fn Status service which can collect run-time status information from SDK Harnesses for
+ * debugging purpose
+ */
+public class BeamWorkerStatusGrpcService extends BeamFnWorkerStatusImplBase implements FnService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamWorkerStatusGrpcService.class);
+  private static final String UNKNOW_SDK_ID_PREFIX = "unknown_sdk";
+  private final AtomicLong idGenerator = new AtomicLong();
+  private final HeaderAccessor headerAccessor;
+  private final Map<String, WorkerStatusClient> connectedClient = new ConcurrentHashMap<>();
+
+  private BeamWorkerStatusGrpcService(
+      ApiServiceDescriptor apiServiceDescriptor, HeaderAccessor headerAccessor) {
+    this.headerAccessor = headerAccessor;
+    LOG.info("Launched Beam Fn Status service at {}", apiServiceDescriptor);
+  }
+
+  /**
+   * Create new instance of {@link BeamWorkerStatusGrpcService}.
+   *
+   * @param apiServiceDescriptor ApiServiceDescriptor used for hosting to the server.
+   * @param headerAccessor Grpc head accessor used to obtain SDK Harness worker id.
+   * @return {@link BeamWorkerStatusGrpcService} if apiServiceDescriptor is valid, otherwise returns
+   *     null.
+   */
+  @Nullable
+  public static BeamWorkerStatusGrpcService create(
+      ApiServiceDescriptor apiServiceDescriptor, HeaderAccessor headerAccessor) {
+    if (apiServiceDescriptor == null || Strings.isNullOrEmpty(apiServiceDescriptor.getUrl())) {
+      LOG.info(
+          "Received empty api service descriptor for status service,"
+              + " the sdk status report will be skipped");
+      return null;
+    }
+    return new BeamWorkerStatusGrpcService(apiServiceDescriptor, headerAccessor);
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (WorkerStatusClient client : this.connectedClient.values()) {
+      client.close();
+    }
+    this.connectedClient.clear();
+  }
+
+  @Override
+  public StreamObserver<WorkerStatusResponse> workerStatus(
+      StreamObserver<WorkerStatusRequest> requestObserver) {
+    String workerId = headerAccessor.getSdkWorkerId();
+    if (Strings.isNullOrEmpty(workerId)) {
+      // SDK harness status report should be enforced with proper sdk worker id.
+      // If SDK harness connect without proper sdk worker id, generate one with name unknown_sdkXX.
+      // TODO: enforce proper worker id.
+      workerId = UNKNOW_SDK_ID_PREFIX + idGenerator.getAndIncrement();
```

Comment:
The related JIRA is closed (BEAM-4149). I don't believe we need this anymore and any existing hack in this space could be deleted as well (in a separate PR).

Reply:
removed.